### DISCUSSION TITLE
bug: can't override gitsigns keymaps with `keys=`

### DISCUSSION THREAD

**Discussion by ilveroluca** [LazyVim/LazyVim#4790] [Q&A]
### Did you check docs and existing issues?

- [x] I have read all the LazyVim docs
- [x] I have updated the plugin to the latest version before submitting this issue
- [x] I have searched the existing issues of LazyVim
- [x] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

v0.10.1

### Operating system/version

Ubuntu 24.04.1

### Describe the bug

I'm trying to override one of the default keymaps for the gitsigns plugin. but the standard `keys = ` mechanism isn't working.

Specifically I'd like to change this map:
```
map("n", "<leader>ghp", gs.preview_hunk_inline, "Preview Hunk Inline")
```
to call `gs.preview_hunk`.  I tried the "normal way", by adding this block to my configuration:
```
{
    "lewis6991/gitsigns.nvim",
    keys = {
        { "<leader>ghp", ":Gitsigns preview_hunk<CR>", desc = "Preview Hunk"},
      }
  },
```
However, it doesn't have any effect.  Same thing if I try to define a new keymap `<leader>ghi` -- it doesn't get picked up either.

### Steps To Reproduce

1. add this block to your plugin configuration:
```
{
    "lewis6991/gitsigns.nvim",
    keys = {
        { "<leader>ghp", ":Gitsigns preview_hunk<CR>", desc = "Preview Hunk"},
      }
  },
```
2. open a file checked out from a local git repository
3. more to a line with a change
4. type the keymap `<leader>ghp`

### Expected Behavior

`:Gitsigns preview_hunk` should be called, instead of the default `Gitsigns preview_hunk_inline`. The difference is that `Gitsigns preview_hunk` opens a dialog showing the diff of the current hunk.

### Repro

```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

require("lazy.minit").repro({
  spec = {
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    -- add any other plugins here
    {
      "lewis6991/gitsigns.nvim",
      keys = {
        { "<leader>ghp", ":Gitsigns preview_hunk<CR>", desc = "Preview Hunk"},
      }
    },
  }
})
```

**Comment by dpetka2001** [comment]
That's not going to work because all the mappings are created in the `on_attach` function of `gitsigns` and they overwrite the `keys` when the plugin loads. Unfortunately, you would have to copy/paste the entire [on_attach function](https://github.com/LazyVim/LazyVim/blob/86b2c23e2aae17e6a887e296e140307e873c8195/lua/lazyvim/plugins/editor.lua#L265-L300) as far as I know, because it's a function and cannot be merged like the rest of the `opts` and you essentially have to overwrite the whole function. 

I might be wrong about it, but this is what I think. If I'm wrong I hope someone corrects me with a more proper solution.

**Comment by gmnz** [comment]
@dpetka2001 What about something like this?

```lua

return {
  "lewis6991/gitsigns.nvim",
  opts = function(_, opts)
    local default_on_attach = opts.on_attach
    opts.on_attach = function(buffer)
      default_on_attach(buffer)
      vim.keymap.set("n", "<leader>ghP", require("gitsigns").preview_hunk, { desc = "Preview Hunk" })
    end
    return opts
  end,
}
```

Anything wrong about solving it this way?

**Reply by dpetka2001** [reply]
Nope. If it works then why not? I just wasn't able to think of this way :stuck_out_tongue: 

Thank you for sharing your solution. You might also want to put `buffer = buffer` to make sure the keymap is only buffer-local when the plugin is attached. And I believe you can also delete buffer-local keymaps with `vim.keymap.del` while making sure to also pass `buffer = buffer`, just so you don't have 2 keymaps with the same action. You can test it out to see if it works and hopefully if it works your answer can be marked as the correct one for others who stumble upon this as well :smile:

**Reply by ilveroluca** [reply]
First of all, thanks for helping!

I originally submitted this as a bug because `keys` in this plugin spec doesn't seem to work [as it should](https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps) -- i.e., *To override a keymap, simply add one with the same lhs and a new rhs.*

@gmnz, your solution for decorating the original `on_attach` to me seems like a good general-purpose strategy and it can work to solve my problem -- thanks!

For my specific question, note that to override the *existing* gitsigns mapping `<leader>ghp` I also need to add the option `buffer = buffer` suggested by @dpetka2001, since the existing gitsigns mappings are all defined on the buffer. Passing `buffer = buffer` in this case is a good idea for new keymaps as well, so that you don't define gitsigns keymaps for buffers where gitsigns isn't attached. I think your example is defining a new global keymap, which would be equivalent to adding it to `config/keymaps.lua`.

My final solution is thus:
```
  {
    "lewis6991/gitsigns.nvim",
    opts = function(_, opts)
      local default_on_attach = opts.on_attach
      opts.on_attach = function(buffer)
        default_on_attach(buffer)
        vim.keymap.set("n", "<leader>ghp", require("gitsigns").preview_hunk, { desc = "Preview Hunk", buffer = buffer })
      end
      return opts
    end,
  },
```

**Comment by mrityunjai01** [comment]
Similar issue, not quite the same plugin. 
In [lazyvim docs](http://www.lazyvim.org/configuration/examples), this keymap to browse plugin files failed to work for me, i believe that telescope here overrides the keymaps, I tried reading the files, e.g. https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/options.lua but couldn't find any mention of `keys`, I have a hard time getting this working using the `keys = ` method as shown here, (taken from [docs](http://www.lazyvim.org/configuration/examples))
```

  -- change some telescope options and a keymap to browse plugin files
  {
    "nvim-telescope/telescope.nvim",
    keys = {
      -- add a keymap to browse plugin files
      -- stylua: ignore
      {
        "<leader>fp",
        function() require("telescope.builtin").find_files({ cwd = require("lazy.core.config").options.root }) end,
        desc = "Find Plugin File",
      },
    },,
```

**Reply by gmnz** [reply]
https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-customizing-plugin-specs
https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps

Does this help?

**Reply by dpetka2001** [reply]
I tried the above code snippet and it works as expected for me. You should put the code snippet in a file in `/lua/plugins/` directory. Keep in mind that the file `example.lua` won't load anything (since it just returns early) and is just meant to be read as an example. So, don't put any customizations in the `example.lua` file. Just create a new file and put your custom settings there.
